diff -up ntfs-3g_ntfsprogs-2014.2.15/src/ntfs-3g_common.c.ref ntfs-3g_ntfsprogs-2014.2.15/src/ntfs-3g_common.c
--- ntfs-3g_ntfsprogs-2014.2.15/src/ntfs-3g_common.c.ref	2015-01-13 14:31:55.648115825 -0500
+++ ntfs-3g_ntfsprogs-2014.2.15/src/ntfs-3g_common.c	2015-01-13 14:32:47.514838649 -0500
@@ -553,7 +553,7 @@ int ntfs_parse_options(struct ntfs_optio
 {
 	int c;
 
-	static const char *sopt = "-o:hnvV";
+	static const char *sopt = "-o:hnsvV";
 	static const struct option lopt[] = {
 		{ "options",	 required_argument,	NULL, 'o' },
 		{ "help",	 no_argument,		NULL, 'h' },
@@ -613,6 +613,11 @@ int ntfs_parse_options(struct ntfs_optio
 			 * we don't use it because mount(8) passes it.
 			 */
 			break;
+		case 's':
+			/*
+			 * no effect - automount passes it, meaning sloppy
+			 */
+			break;
 		case 'V':
 			ntfs_log_info("%s %s %s %d\n", EXEC_NAME, VERSION, 
 				      FUSE_TYPE, fuse_version());
